main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+      *
+      * Examples:
+      * -     true ~ x7 ==> x7 ~ true
+      * -       c3 ~ c2 ==> c2 ~ c3
+      * -       x7 ~ x5 ==> x5 ~ x7
+      * - x3 /\ x7 ~ x4 ==> x4 ~ x3 /\ x7
+      * - x1 /\ x2 ~ c5 ==> c5 ~ x1 /\ x2
+      */
+    def mk(t1: Term, t2: Term, loc: SourceLocation): Equation = (t1, t2) match {
+      case (Term.Cst(c1), Term.Cst(c2)) => if (c1 <= c2) Equation(t1, t2, loc) else Equation(t2, t1, loc)
+      case (Term.Var(x1), Term.Var(x2)) => if (x1 <= x2) Equation(t1, t2, loc) else Equation(t2, t1, loc)
+      case (Term.True, _) => Equation(t2, Term.True, loc)
+      case (Term.False, _) => Equation(t2, Term.False, loc)
+      case (_, Term.Var(_)) => Equation(t2, t1, loc)
+      case _ => Equation(t1, t2, loc)
+    }
+  }
+
+  /**
+    * Represents a unification equation `t1 ~ t2` between the terms `t1` and `t2`.
+    *
+    * WARNING: Equations should be normalized. Use the smart constructor [[Equation.mk]] to create a new equation.
+    */
+  case class Equation(t1: Term, t2: Term, loc: SourceLocation) {
+    /**
+      * Returns the size of this equation which is the sum of its lhs and rhs.
+      */
+    final def size: Int = t1.size + t2.size
+
+    /**
+      * Returns a human-readable representation of `this` unification equation.
+      */
+    override final def toString: String = s"$t1 ~ $t2"
+  }
+
+  /**
+    * A common super-type for Boolean terms.
+    *
+    * Note: We assume that the integers used for constants and variables are disjoint.
+    */
+  sealed trait Term {
+
+    /**
+      * Syntactic sugar for [[Term.mkAnd]].
+      */
+    final def &(that: Term): Term = Term.mkAnd(this, that)
+
+    /**
+      * Syntactic sugar for [[Equation.mk]]
+      */
+    final def ~(that: Term)(implicit loc: SourceLocation): Equation = Equation.mk(this, that, loc)
+
+    /**
+      * Returns all variables that occur in `this` term.
+      */
+    final def freeVars: SortedSet[Int] = this match {
+      case Term.True => SortedSet.empty
+      case Term.False => SortedSet.empty
+      case Term.Cst(_) => SortedSet.empty
+      case Term.Var(x) => SortedSet(x)
+      case Term.Not(t) => t.freeVars
+
+      case Term.And(_, vars, rest) => SortedSet.empty[Int] ++ vars.map(_.x) ++ rest.flatMap(_.freeVars)
+
+      case Term.Or(ts) => ts.foldLeft(SortedSet.empty[Int])(_ ++ _.freeVars)
+    }
+
+    /**
+      * Returns the size of `this` term.
+      *
+      * The size is the number of constants, variables, and connectives in the term.
+      *
+      * For example, `size(x /\ y) = 1` and `size(x /\ not y) = 2`.
+      */
+    final def size: Int = this match {
+      case Term.True => 0
+      case Term.False => 0
+      case Term.Cst(_) => 1
+      case Term.Var(_) => 0
+      case Term.Not(t) => t.size + 1
+      case Term.And(csts, vars, rest) =>
+        // We need a connective for each constant, variable, and term minus one.
```

Comment:
This does not seem consistent. A `Cst` `X` has size 1. I would then expect `X and X` to have size 3, but it looks like it is 1 here.

Reply:
I changed it so we only count connectives.